src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java

```diff
@@ -226,6 +235,95 @@ private void initializeNucleusComponentConfig() {
+    private void persistInitialLaunchParams(KernelAlternatives kernelAlts) {
+        // Persist initial Nucleus launch parameters
+        try {
+            String jvmOptions = ManagementFactory.getRuntimeMXBean().getInputArguments().stream().sorted()
+                    .filter(s -> !s.startsWith(JVM_OPTION_ROOT_PATH)).collect(Collectors.joining(" "));
+            kernel.getConfig().lookup(SERVICES_NAMESPACE_TOPIC, getNucleusComponentName(), CONFIGURATION_CONFIG_KEY,
+                    DEVICE_PARAM_JVM_OPTIONS).dflt(jvmOptions);
+
+            kernelAlts.writeLaunchParamsToFile(jvmOptions);
+            logger.atInfo().log("Successfully setup Nucleus launch parameters");
+        } catch (IOException e) {
+            logger.atError().log("Unable to setup Nucleus launch parameters", e);
+        }
+    }
+
+    /**
+     * Load Nucleus component information from build recipe.
+     *
+     * @param kernelAlts KernelAlternatives instance
+     */
+    public void initializedNucleusFromRecipe(KernelAlternatives kernelAlts) {
+        String nucleusComponentName = getNucleusComponentName();
+        Topics nucleusLifecycle = kernel.getConfig().lookupTopics(DEFAULT_VALUE_TIMESTAMP, SERVICES_NAMESPACE_TOPIC,
+                nucleusComponentName, SERVICE_LIFECYCLE_NAMESPACE_TOPIC);
+        if (nucleusLifecycle.children.size() != 0) {
+            logger.atDebug().log("Nucleus lifecycle has already been initialized");
+            return;
+        }
+
+        persistInitialLaunchParams(kernelAlts);
+        try {
+            Path unpackDir = locateCurrentKernelUnpackDir();
+            Path recipePath = unpackDir.resolve(NUCLEUS_BUILD_METADATA_DIRECTORY)
+                    .resolve(NUCLEUS_RECIPE_FILENAME);
+
+            // Copy recipe to component store
+            ComponentStore componentStore = kernel.getContext().get(ComponentStore.class);
+            DeploymentService.copyRecipeFileToComponentStore(componentStore, recipePath, logger);
+            // Copy artifact to component store
+            // TODO
+
+            // Update Nucleus in config store
+            Optional<ComponentRecipe> resolvedRecipe = kernel.getContext().get(RecipeLoader.class)
+                    .loadFromFile(new String(Files.readAllBytes(recipePath.toAbsolutePath()), StandardCharsets.UTF_8));
+            if (!resolvedRecipe.isPresent()) {
+                throw new PackageLoadingException("Failed to resolve Nucleus recipe");
+            }
+            ComponentRecipe componentRecipe = resolvedRecipe.get();
+            nucleusComponentName = componentRecipe.getComponentName();
+
+            // Add Nucleus version
+            String componentVersion = componentRecipe.getVersion().toString();
+            kernel.getConfig().lookup(SERVICES_NAMESPACE_TOPIC, nucleusComponentName,
+                    VERSION_CONFIG_KEY).dflt(componentVersion);
+
+            kernel.getConfig().lookup(SETENV_CONFIG_NAMESPACE, GGC_VERSION_ENV).withValue(componentVersion);
```

Comment:
you can't put this here since in most cases this code will never execute. Put this somewhere that always executes so that we always pass the version in.

Reply:
Ok. I can move it back to where it was